File: backend/app/schemas/energy_certificate.py
"""
Energy Certificate Schemas
"""
from typing import Optional
from datetime import date
from pydantic import BaseModel, Field


class EnergyCertificateUpdate(BaseModel):
    """Schema for updating energy certificate data"""
    energy_class: Optional[str] = Field(None, description="Energy efficiency class (A+ to H)")
    energy_consumption: Optional[int] = Field(None, description="Energy consumption in kWh/m²a")
    energy_certificate_type: Optional[str] = Field(None, description="Certificate type")
    energy_certificate_valid_until: Optional[date] = Field(None, description="Certificate valid until date")
    energy_certificate_issue_date: Optional[date] = Field(None, description="Certificate issue date")
    co2_emissions: Optional[int] = Field(None, description="CO₂ emissions in kg/m²a")
    heating_type: Optional[str] = Field(None, description="Heating type")


class EnergyCertificateResponse(BaseModel):
    """Schema for energy certificate response"""
    energy_class: Optional[str] = None
    energy_consumption: Optional[int] = None
    energy_certificate_type: Optional[str] = None
    energy_certificate_valid_until: Optional[date] = None
    energy_certificate_issue_date: Optional[date] = None
    co2_emissions: Optional[int] = None
    heating_type: Optional[str] = None


class EnergyCertificatePDFRequest(BaseModel):
    """Schema for PDF generation request"""
    include_logo: bool = Field(True, description="Include company logo in PDF")
    language: str = Field("de", description="PDF language (de/en)")
    template: str = Field("standard", description="PDF template to use")


class EnergyCertificatePDFResponse(BaseModel):
    """Schema for PDF generation response"""
    pdf_url: str = Field(..., description="URL to download the generated PDF")
    filename: str = Field(..., description="Suggested filename for download")
    generated_at: str = Field(..., description="Generation timestamp")

File: frontend/src/services/energyCertificate.ts
/**
 * Energy Certificate Service
 * API calls for energy certificate management
 */

import { apiClient } from '../api/config';

export interface EnergyCertificateData {
  energy_class?: string;
  energy_consumption?: number;
  energy_certificate_type?: string;
  energy_certificate_valid_until?: string;
  energy_certificate_issue_date?: string;
  co2_emissions?: number;
  heating_type?: string;
}

export interface EnergyCertificateUpdateRequest {
  energy_class?: string;
  energy_consumption?: number;
  energy_certificate_type?: string;
  energy_certificate_valid_until?: string;
  energy_certificate_issue_date?: string;
  co2_emissions?: number;
  heating_type?: string;
}

export interface EnergyCertificatePDFRequest {
  include_logo?: boolean;
  language?: string;
  template?: string;
}

export interface EnergyCertificatePDFResponse {
  pdf_url: string;
  filename: string;
  generated_at: string;
}

export class EnergyCertificateService {
  /**
   * Get energy certificate data for a property
   */
  static async getEnergyData(propertyId: string): Promise<EnergyCertificateData> {
    try {
      const response = await apiClient.get<EnergyCertificateData>(
        `/api/v1/properties/${propertyId}/energy-data`
      );
      return response.data;
    } catch (error: any) {
      console.error('Error fetching energy data:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Laden der Energiedaten');
    }
  }

  /**
   * Update energy certificate data for a property
   */
  static async updateEnergyData(
    propertyId: string,
    data: EnergyCertificateUpdateRequest
  ): Promise<EnergyCertificateData> {
    try {
      const response = await apiClient.put<EnergyCertificateData>(
        `/api/v1/properties/${propertyId}/energy-data`,
        data
      );
      return response.data;
    } catch (error: any) {
      console.error('Error updating energy data:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Speichern der Energiedaten');
    }
  }

  /**
   * Generate energy certificate PDF
   */
  static async generatePDF(
    propertyId: string,
    request: EnergyCertificatePDFRequest = {}
  ): Promise<EnergyCertificatePDFResponse> {
    try {
      const response = await apiClient.post<EnergyCertificatePDFResponse>(
        `/api/v1/properties/${propertyId}/energy-certificate/generate`,
        {
          include_logo: true,
          language: 'de',
          template: 'standard',
          ...request
        }
      );
      return response.data;
    } catch (error: any) {
      console.error('Error generating PDF:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Generieren des PDFs');
    }
  }

  /**
   * Download energy certificate PDF
   */
  static async downloadPDF(propertyId: string): Promise<Blob> {
    try {
      const response = await apiClient.get(
        `/api/v1/properties/${propertyId}/energy-certificate/download`,
        {
          responseType: 'blob'
        }
      );
      return response.data as Blob;
    } catch (error: any) {
      console.error('Error downloading PDF:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Herunterladen des PDFs');
    }
  }

  /**
   * Download PDF and trigger browser download
   */
  static async downloadPDFFile(propertyId: string, filename?: string): Promise<void> {
    try {
      const blob = await this.downloadPDF(propertyId);
      
      // Create download link
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.download = filename || `Energieausweis_${propertyId}_${new Date().toISOString().split('T')[0]}.pdf`;
      
      // Trigger download
      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);
      
      // Clean up
      window.URL.revokeObjectURL(url);
    } catch (error: any) {
      console.error('Error downloading PDF file:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Herunterladen des PDFs');
    }
  }
}
